File: Sorting/Python/selectionSort.py
def selectionSort(arr):
    n = len(arr)
    # Find the minimum index first and then swap it with the current index
    for i in range(n-1):
        # Assume the current index is the minimum
        mini = i
        # Find the index of the minimum element in the unsorted portion
        for j in range(i+1, n):
            if arr[j] < arr[mini]:
                mini = j
        # Swap the minimum element with the first element of the unsorted portion
        arr[mini], arr[i] = arr[i], arr[mini]
    return arr

# The best, worst, and average time complexities for the selection sort is: O(N*N)

def test_selection_sort():
    # Test Case 1: Random array
    arr1 = [54, 26, 93, 17, 77, 31, 44, 55, 20]
    assert selectionSort(arr1) == sorted(arr1)

    # Test Case 2: Sorted array (ascending)
    arr2 = [10, 20, 30, 40, 50, 60]
    assert selectionSort(arr2) == sorted(arr2)

    # Test Case 3: Sorted array (descending)
    arr3 = [90, 80, 70, 60, 50, 40]
    assert selectionSort(arr3) == sorted(arr3)

    # Test Case 4: Array with duplicate elements
    arr4 = [5, 2, 7, 2, 3, 5, 9, 3]
    assert selectionSort(arr4) == sorted(arr4)

    # Test Case 5: Array with negative numbers
    arr5 = [-10, -5, -3, -8, -15]
    assert selectionSort(arr5) == sorted(arr5)

    # Test Case 6: Empty array
    arr6 = []
    assert selectionSort(arr6) == sorted(arr6)

    # Test Case 7: Array with one element
    arr7 = [42]
    assert selectionSort(arr7) == sorted(arr7)

# Run the test cases
test_selection_sort()
print("All test cases passed!")

File: Sorting/Java/SelectionSort.java
package Sorting.Java;

public class SelectionSort {
    // Selecting the minimum and swapping it with the first unsorted element
    public static void selectionSorted(int[] arr) {
        for(int i=0;i<arr.length;i++){
            int minIndex = i; // Assume the first element is the minimum
            for(int j=i+1;j<arr.length;j++){
                if(arr[j]<arr[minIndex]){
                    minIndex = j; // Update minIndex if a smaller element is found
                }
            }
            if(minIndex != i) swap(arr, i, minIndex); // Swap the found minimum element with the first unsorted element
        }
    }

    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public static void main(String[] args) {
        int[] arr = {38,15,26,4,43,8,19,37,49,2};
        selectionSorted(arr);
        System.out.print("Sorted array: ");
        for(int i: arr)  System.out.print(i + " ");
        System.out.println();
    }
}

// Time Complexity: O(n^2) in all cases (best, average, worst)
// Space Complexity: O(1) as it sorts in place
// This implementation of Selection Sort is stable and sorts the array in ascending order.
